### PR TITLE
ARROW-10150: [C++] Fix crashes on invalid Parquet file

### PR DESCRIPTION
**Pull request by pitrou**
Found by OSS-Fuzz. Should fix the following issues:
- https://bugs.chromium.org/p/oss-fuzz/issues/detail?id=26064
- https://bugs.chromium.org/p/oss-fuzz/issues/detail?id=26065
- https://bugs.chromium.org/p/oss-fuzz/issues/detail?id=26067

### CODE REVIEW

File: cpp/src/parquet/arrow/schema.cc [hunk 1]
```diff
@@ -693,10 +693,12 @@ Status ApplyOriginalStorageMetadata(const Field& origin_field, SchemaField* infe
...
 
-  if (inferred_type->id() == ::arrow::Type::TIMESTAMP) {
+  if (origin_type->id() == ::arrow::Type::TIMESTAMP &&
+      inferred_type->id() == ::arrow::Type::TIMESTAMP) {
     // Restore time zone, if any
-    const auto& ts_type = static_cast<const ::arrow::TimestampType&>(*inferred_type);
-    const auto& ts_origin_type = static_cast<const ::arrow::TimestampType&>(*origin_type);
+    const auto& ts_type = checked_cast<const ::arrow::TimestampType&>(*inferred_type);
+    const auto& ts_origin_type =
+        checked_cast<const ::arrow::TimestampType&>(*origin_type);
```
Comment (emkornfield) [inline cpp/src/parquet/arrow/schema.cc:701]: I thought checked_cast only had functionality in debug most and was essentially a static cast in prod?  Or is this not the root cause of the issue?

**Review by emkornfield — APPROVED** [review]
It seems like checked_cast is only for style?

File: cpp/src/parquet/arrow/schema.cc (hunk 1 above)
Comment (pitrou) [reply cpp/src/parquet/arrow/schema.cc:701]: Well, it's useful to have a debug mode check, IMHO.

**Comment by emkornfield** [comment]
+1 merging.

**Pull request closed without merging**